src/crimson/osd/shard_services.cc

```diff
@@ -0,0 +1,223 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/osd/shard_services.h"
+#include "messages/MOSDPGTemp.h"
+#include "messages/MOSDPGCreated.h"
+#include "messages/MOSDPGNotify.h"
+#include "messages/MOSDPGInfo.h"
+#include "messages/MOSDPGQuery.h"
+
+namespace {
+  seastar::logger& logger() {
+    return ceph::get_logger(ceph_subsys_osd);
+  }
+}
+
+namespace ceph::osd {
+
+ShardServices::ShardServices(
+  ceph::net::Messenger &cluster_msgr,
+  ceph::net::Messenger &public_msgr,
+  ceph::mon::Client &monc,
+  ceph::mgr::Client &mgrc,
+  ceph::os::CyanStore &store)
+    : cluster_msgr(cluster_msgr),
+      public_msgr(public_msgr),
+      monc(monc),
+      mgrc(mgrc),
+      store(store) {
+  perf = build_osd_logger(&cct);
+  cct.get_perfcounters_collection()->add(perf);
+
+  recoverystate_perf = build_recoverystate_perf(&cct);
+  cct.get_perfcounters_collection()->add(recoverystate_perf);
+}
+
+seastar::future<> ShardServices::send_to_osd(
+  int peer, Ref<Message> m, epoch_t from_epoch) {
+  if (osdmap->is_down(peer) || osdmap->get_info(peer).up_from > from_epoch) {
+    return seastar::now();
+  } else {
+    return cluster_msgr.connect(osdmap->get_cluster_addrs(peer).front(),
+      CEPH_ENTITY_TYPE_OSD)
+      .then([m, this] (auto xconn) {
+	      return (*xconn)->send(m);
+	    });
+  }
+}
+
+seastar::future<> ShardServices::dispatch_context_transaction(
+  ceph::os::CyanStore::CollectionRef col, PeeringState::PeeringCtx &ctx) {
+  auto ret = store.do_transaction(
+    col,
+    std::move(ctx.transaction));
+  ctx.reset_transaction();
+  return ret;
+}
+
+seastar::future<> ShardServices::dispatch_context_messages(
+  PeeringState::PeeringCtx &ctx)
+{
+  auto ret = seastar::do_with(PeeringState::PeeringCtx{ctx}, [this](auto& todo) {
```

Comment:
can we capture-by-move `ctx.notify_list`, `ctx.query_map` and `ctx.info_map` instead of copying them?

Reply:
I think I did for the bits below, but I'll update the initial do_with to capture-by-move PeeringCtx.